pkg/v0/cmd.go

```diff
@@ -160,9 +171,127 @@ func Run(ctx context.Context, logger *logrus.Logger, opts Options) error {
 	return nil
 }
 
+func getTagOrCommit(ctx context.Context, repo string, dir string, opts Options, logger *logrus.Entry) (string, error) {
+
+	// Create temporary
+
+	module := fmt.Sprintf("github.com/%s", repo)
+	rawInfo, err := internal.RunCommand(logger, internal.WithDir(exec.CommandContext(ctx,
+		"go", "list", "-json", "-m", module), dir))
+	if err != nil {
+		return "", fmt.Errorf("failed to determine dependent version for module %s: %w", module, err)
+	}
+	var info struct {
+		Version string `json:"Version"`
+	}
+	if  err := json.Unmarshal([]byte(rawInfo), &info); err != nil {
+		return "", fmt.Errorf("failed to parse module version for %s: %w", module, err)
+	}
+	logger.WithFields(logrus.Fields{"repo": repo, "version": info.Version}).Info("resolved latest version")
+
+	v, err := semver.NewVersion(info.Version)
+	if err != nil {
+		return "", err
+	}
+	// If this does not have a Prerelease, then we just return the version string
+	pre := v.Prerelease()
+	if pre == "" {
+		return info.Version, nil
+	}
+	// It's a pre-release version, which we assume is in DATE-COMMIT format
+	pres := strings.Split(pre, "-")
+	if len(pres) != 2 {
+		return "", fmt.Errorf("Bad prerelease: %q", info.Version)
+	}
+	// Return the second component, which is a commit SHA
+	return pres[1], nil
+}
+
+func calculateRepoRefs(ctx context.Context, logger *logrus.Entry, opts Options) (map[string]string, error) {
+	repoRefs := map[string]string{}
+
+	// for operator-lifecycle-manager, always use master
+	repoRefs["operator-framework/operator-lifecycle-manager"] = "master"
+
+	// Create a temporary worktree of upstream OLM to figure out what dependency versions we are moving to
+	var remote string
+
+	switch flags.FetchMode(opts.FetchMode) {
+	case flags.SSH:
+		remote = "git@github.com:operator-framework/operator-lifecycle-manager"
+	case flags.HTTPS:
+		remote = "https://github.com/operator-framework/operator-lifecycle-manager.git"
+	}
+	if _, err := internal.RunCommand(logger, exec.CommandContext(ctx,
+		"git", "fetch",
+		remote,
+		"master",
+	)); err != nil {
+		return nil, err
+	}
+	dir, err := os.MkdirTemp("", "olm")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(dir)
+	if _, err := internal.RunCommand(logger, exec.CommandContext(ctx,
+		"git", "worktree",
+		"add",
+		dir,
+		"FETCH_HEAD",
+	)); err != nil {
+		return nil, err
+	}
```

Comment:
TIL about `git worktree` - neat!

Do we need to be concerned about any branches this might create on a local machine? Reading the [documentation on `git worktree`](https://git-scm.com/docs/git-worktree) it looks like we could use the `-d` flag to create a "throwaway" worktree (although maybe this doesn't matter at all since we attempt to clean up the directory later).

Reply:
Yes, and the neat thing is that it can be _any_ reference in the repo, including things you fetched from outside of the origin.
When run as a job, it's done in a temporary container, so cleanup isn't critical.
_But_ if you run it locally, it will still show up in `git worktree list`, and using the `git worktree prune` will cleanup any number of them. In my case, I had many left over from testing.

Reply:
I would recommend not doing this. We have many years of experience with automated tooling that uses `git` and not storing references in filesystem state (branches, etc) is a less error-prone approach. That way, your program has no expectations about the environment it's run in and it can be correct regardless of what kind of `git` state exists in a container, on a dev machine, whatever. Plus, no cleanup necessary.